### PR TITLE
[expo-segment] Add Segment.optOut feature

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

Overrides, closes https://github.com/expo/expo/pull/1004.

Also, https://expo.canny.io/feature-requests/p/add-a-closepause-method-to-segment. 

# How

- Cherry-picked @ctgardner's commit, moved the changes to `expo-analytics-admob`.
- Added setting persistence across restarts.
- Added promise rejection for trying to set enabled state in Expo Client.

# Test Plan

`test-suite` tests included, passing on iOS and Android.

### CODE REVIEW

File: packages/expo-analytics-segment/src/Segment.js [hunk 1]
```diff
@@ -55,4 +55,12 @@ export default {
   flush(): void {
     ExponentSegment.flush();
   },
+
+  getEnabledAsync(): Promise<boolean> {
+    return ExponentSegment.getEnabledAsync();
+  },
+
+  setEnabledAsync(enabled: boolean): Promise<void> {
+    return ExponentSegment.setEnabledAsync(true);
```
Comment (tsapeta) [inline packages/expo-analytics-segment/src/Segment.js]: ExponentSegment.setEnabledAsync(~true~ **enabled**)?

File: packages/expo-analytics-segment/ios/EXSegment.podspec [hunk 2]
```diff
@@ -17,6 +17,7 @@ Pod::Spec.new do |s|
   s.requires_arc   = true
 
   s.dependency 'EXCore'
+  s.dependency 'EXConstantsInterface'
```
Comment (tsapeta) [inline packages/expo-analytics-segment/ios/EXSegment.podspec]: Looks like there is no `peerDependency` on `expo-constants-interface`

File: packages/expo-analytics-segment/ios/EXSegment/EXSegment.m [hunk 3]
```diff
@@ -1,19 +1,26 @@
 // Copyright 2015-present 650 Industries. All rights reserved.
 
 #import <EXSegment/EXSegment.h>
-
+#import <EXConstantsInterface/EXConstantsInterface.h>
 #import <SEGAnalytics.h>
 
+static NSString * const EXSegmentOptOutKey = @"EXSegmentOptOutKey";
```
Comment (tsapeta) [inline packages/expo-analytics-segment/ios/EXSegment/EXSegment.m]: I think we used to be doing `NSString *const`

File: packages/expo-analytics-segment/ios/EXSegment/EXSegment.m [hunk 4]
```diff
@@ -146,4 +157,32 @@ @implementation EXSegment
   resolve(nil);
 }
 
+EX_EXPORT_METHOD_AS(getEnabledAsync,
+                    getEnabledWithResolver:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  NSNumber *optOutSetting = [[NSUserDefaults standardUserDefaults] objectForKey:EXSegmentOptOutKey];
+  resolve(optOutSetting ?: @(true));
```
Comment (tsapeta) [inline packages/expo-analytics-segment/ios/EXSegment/EXSegment.m]: `true` -> `YES`

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: ios/Podfile.lock [hunk 5]
```diff
@@ -1384,7 +1385,7 @@ SPEC CHECKSUMS:
   EXPermissionsInterface: 76f80a0dcd16afde36498e3d3aa168cb43999558
   EXPrint: b6aab9715518cf67954ae260b71e28afd843353e
   EXReactNativeAdapter: 253efa069b53a6c8be7c3f2afd49ef6b4e97e902
-  EXSegment: b39f68ecbd4f7262b1f9cb3f187e684c92651894
+  EXSegment: cfe9aed31dbd7f1455724afe89a4920e5dad0049
```
Comment (bbarthec) [inline ios/Podfile.lock]: Well, I see these two changes in this file, but is it possible that these don't affect file final checksum? 🤔

File: packages/expo-analytics-segment/android/src/main/java/expo/modules/analytics/segment/SegmentModule.java [hunk 6]
```diff
@@ -17,21 +18,34 @@
...
+import expo.core.ModuleRegistry;
 import expo.core.Promise;
 import expo.core.interfaces.ExpoMethod;
+import expo.core.interfaces.ModuleRegistryConsumer;
+import expo.interfaces.constants.ConstantsInterface;
 
-public class SegmentModule extends ExportedModule {
-
+public class SegmentModule extends ExportedModule implements ModuleRegistryConsumer {
+  private static final String NAME = "ExponentSegment";
```
Comment (bbarthec) [inline packages/expo-analytics-segment/android/src/main/java/expo/modules/analytics/segment/SegmentModule.java]: Well, `ExponentSegment` -> `ExpoSegment`?

File: packages/expo-analytics-segment/ios/EXSegment/EXSegment.m [hunk 7]
```diff
@@ -1,19 +1,26 @@
...
 @interface EXSegment ()
 
 @property (nonatomic, strong) SEGAnalytics *instance;
+@property (nonatomic, weak) id<EXConstantsInterface> constants;
 
 @end
 
 @implementation EXSegment
 
 EX_EXPORT_MODULE(ExponentSegment)
```
Comment (bbarthec) [inline packages/expo-analytics-segment/ios/EXSegment/EXSegment.m]: `Exponent` -> `Expo`

File: packages/expo-analytics-segment/ios/EXSegment/EXSegment.m [hunk 8]
```diff
@@ -1,19 +1,26 @@
...
 
+- (void)setModuleRegistry:(EXModuleRegistry *)moduleRegistry
+{
+  _constants = [moduleRegistry getModuleImplementingProtocol:@protocol(EXConstantsInterface)];
```
Comment (bbarthec) [inline packages/expo-analytics-segment/ios/EXSegment/EXSegment.m]: I believe we should introduce some `assertions` for retrieving modules from registry. What you say?

File: packages/expo-analytics-segment/src/Segment.js [hunk 9]
```diff
@@ -55,4 +55,12 @@ export default {
...
+  },
```
Comment (bbarthec) [inline packages/expo-analytics-segment/src/Segment.js]: Would you mind transforming that into `typescript`?

**Review by tsapeta — APPROVED** [review]

**Pull request merged**